xios: update to v2.6

LaurentChardon opened the pull request:
- xios: update to v2.6
- xios: patch remap earcut.hpp for missing include file
- xios: address C++11 requirement
- xios: instruct to use external boost and blitz libraries

I couldn't compile any of the existing versions of xios. So I fetched the latest version and made it work. I've left the old, non-working versions in place, but I can remove them if required. Since there are no dependants on xios in spack, it might be safe to remove the old versions.

LaurentChardon committed:
xios: update to v2.6

- xios: update to v2.6
- xios: patch remap earcut.hpp for missing include file
- xios: address C++11 requirement
- xios: instruct to use external boost and blitz libraries

LaurentChardon committed:
Merge branch 'spack:develop' into develop

LaurentChardon committed:
Merge branch 'spack:develop' into develop

LaurentChardon committed:
Merge branch 'spack:develop' into develop

LaurentChardon committed:
Merge branch 'spack:develop' into develop

wdconinc commented on var/spack/repos/builtin/packages/xios/package.py:
```diff
@@ -40,6 +45,7 @@ class Xios(Package):
     # followed by a character is broken (e.g. duration '1d'):
     patch("llvm_bug_17782.patch", when="@1.1: %apple-clang")
     patch("llvm_bug_17782.patch", when="@1.1: %clang")
```
(Optional) Consider applying this patch regardless of compiler. There are many llvm-based compilers that would need to be listed. If the patch results in valid code for all compilers, then it should be fine to apply always. It also helps when checking whether the patch applies cleanly when adding new version. Which leads me to...

Did you check that this patch still applies for 2.6?

LaurentChardon replied:
Thanks for reviewing my PR @wdconinc.
I did not check these patches because I was not using clang. I will test to see if they are still necessary with clang.

LaurentChardon replied:
I looked at the code and the patch is for a clang compiler bug that's still unresolved (https://bugs.llvm.org/show_bug.cgi?id=17782). It's not good to apply this patch to compilers other than clang because, even though the patch is necessary for clang, the code is worse with the patch, not better. I'm leaving the patch as it is and I've added a comment that it was checked today.

wdconinc reviewed (commented):
Thanks for the update.

wdconinc commented on var/spack/repos/builtin/packages/xios/package.py:
```diff
@@ -12,10 +12,15 @@ class Xios(Package):
+        revision=2703,
+        svn="https://forge.ipsl.jussieu.fr/ioserver/svn/XIOS2/branches/xios-2.6",
+    )
+    version(
+        "2.5",
+        revision=1860,
+        svn="https://forge.ipsl.jussieu.fr/ioserver/svn/XIOS2/branches/xios-2.5",
     )
     version(
         "2.0", revision=1627, svn="http://forge.ipsl.jussieu.fr/ioserver/svn/XIOS/branchs/xios-2.0"
```
Older versions can be marked as deprecated, and they can get removed after the next release. We typically avoid simply removing a version (even if you cannot verify that it compiles correctly since someone else on a different system may still be using the older version).

E.g.
```suggestion
    version(
        "2.5",
        revision=1860,
        svn="https://forge.ipsl.jussieu.fr/ioserver/svn/XIOS2/branches/xios-2.5",
        deprecated=True,
    )
    version(
        "2.0", revision=1627, svn="http://forge.ipsl.jussieu.fr/ioserver/svn/XIOS/branchs/xios-2.0"
```

LaurentChardon replied:
I'll deprecate the older versions this way. Thanks for the suggestion.

LaurentChardon replied:
Done

LaurentChardon committed:
Merge branch 'spack:develop' into develop

LaurentChardon committed:
xios: bump revision to 2714

- Bump revision to changeset 2714
- Deprecate older versions that I can't manage to compile
- Confirm clang workaround is still needed
